### PR TITLE
Modify demo_periodic_gep.py so that it also demonstrates the use of t…

### PR DESCRIPTION
**Pull request by fmonteghetti**
This pull request addresses Issue #21 by modifying demo_periodic_gep.py to demonstrate a workaround.

The script has been tested with both single and multiple threads in docker image dokken92/dolfinx_mpc:v0.4.1.

The modifications are:
   - The script now also demonstrate the use of two overlapping periodic boundary conditions. The slave_to_master map of the second periodic boundary condition is patched just before calling `create_periodic_constraint_topological()`. My opinion is that separating the patching from the initial definition of the slave_to_master map (which follows the mathematical problem) makes the script clearer and suggests a natural generalization to N overlapping periodic boundary conditions.
   - Comparison to exact eigenvalues have been simplified.

### CODE REVIEW

File: python/demos/demo_periodic_gep.py [hunk 1]
```diff
@@ -198,82 +205,143 @@ def solve_GEP_shiftinvert(A: PETSc.Mat, B: PETSc.Mat,
...
+    boundary_condition
+        First item describes b.c. on {x=0} and {x=1}
+        Second item describes b.c. on {y=0} and {y=1}
+    Nev
+        Number of requested eigenvalues. The default is 10.
+    """
+    comm = MPI.COMM_WORLD
+    # Create mesh and finite element
+    N = 50
+    mesh = create_unit_square(MPI.COMM_WORLD, N, N)
```
Comment (jorgensd) [inline python/demos/demo_periodic_gep.py:225]:
```suggestion
    comm = MPI.COMM_WORLD
    # Create mesh and finite element
    N = 50
    mesh = create_unit_square(comm, N, N)
```

File: python/demos/demo_periodic_gep.py [hunk 2]
```diff
@@ -198,82 +205,143 @@ def solve_GEP_shiftinvert(A: PETSc.Mat, B: PETSc.Mat,
...
+    V = fem.FunctionSpace(mesh, ("CG", 1))
```
Comment (jorgensd) [inline python/demos/demo_periodic_gep.py:226]:
```suggestion
    V = fem.FunctionSpace(mesh, ("Lagrange", 1))
```

File: python/demos/demo_periodic_gep.py [hunk 3]
```diff
@@ -198,82 +205,143 @@ def solve_GEP_shiftinvert(A: PETSc.Mat, B: PETSc.Mat,
...
+    fdim = mesh.topology.dim - 1
+    
+    bcs = []
+    # Dirichlet boundary condition on {y=0} and {y=1}
+    if boundary_condition[1]=="dirichlet":
```
Comment (jorgensd) [inline python/demos/demo_periodic_gep.py:231]:
```suggestion
    if boundary_condition[1] == "dirichlet":
```

File: python/demos/demo_periodic_gep.py [hunk 4]
```diff
@@ -198,82 +205,143 @@ def solve_GEP_shiftinvert(A: PETSc.Mat, B: PETSc.Mat,
...
+        u_bc = fem.Function(V)
+        with u_bc.vector.localForm() as u_local:
+            u_local.set(0.0)
```
Comment (jorgensd) [inline python/demos/demo_periodic_gep.py:234]:
```suggestion
        u_bc.x.array[:]  =0
```

File: python/demos/demo_periodic_gep.py [hunk 5]
```diff
@@ -198,82 +205,143 @@ def solve_GEP_shiftinvert(A: PETSc.Mat, B: PETSc.Mat,
...
+            out_x[1] = x[1] - 1
+            out_x[2] = x[2]
+            return out_x
+        facets = locate_entities_boundary(mesh, fdim, pbc_vert_is_slave)
+        arg_sort = np.argsort(facets)
+        pbc_vert_mt = meshtags(mesh, fdim, facets[arg_sort], np.full(len(facets), pbc_vert_slave_tag, dtype=np.int32))
+        
+    # Create MultiPointConstraint object
+    mpc = MultiPointConstraint(V)
+    if boundary_condition[0]=="periodic":
```
Comment (jorgensd) [inline python/demos/demo_periodic_gep.py:273]: Make sure that you run flake8 on the demo

File: python/demos/demo_periodic_gep.py [hunk 6]
```diff
@@ -198,82 +205,143 @@ def solve_GEP_shiftinvert(A: PETSc.Mat, B: PETSc.Mat,
...
+                                                   bcs)
+    mpc.finalize()
+    # Define variational problem
+    u = TrialFunction(V)
+    v = TestFunction(V)
+    a = inner(grad(u), grad(v)) * dx
+    b = inner(u, v) * dx
+    mass_form = fem.form(a)
+    stiffness_form = fem.form(b)
+    # Diagonal values for slave and Dirichlet DoF
```
Comment (jorgensd) [inline python/demos/demo_periodic_gep.py:300]:
```suggestion

    # Diagonal values for slave and Dirichlet DoF
```

File: python/demos/demo_periodic_gep.py [hunk 7]
```diff
@@ -198,82 +205,143 @@ def solve_GEP_shiftinvert(A: PETSc.Mat, B: PETSc.Mat,
...
+    diagval_A = 1e2
+    diagval_B = 1e-2
+    A = assemble_matrix(mass_form, mpc, bcs=bcs, diagval=diagval_A)
+    B = assemble_matrix(stiffness_form, mpc, bcs=bcs, diagval=diagval_B)
+    EPS = solve_GEP_shiftinvert(A, B, problem_type=SLEPc.EPS.ProblemType.GHEP,
+                                solver=SLEPc.EPS.Type.KRYLOVSCHUR,
+                                nev=Nev, tol=1e-7, max_it=10,
+                                target=1.5, shift=1.5,comm=comm)
+    (eigval, eigvec_r, eigvec_i) = EPS_get_spectrum(EPS, mpc)
+    # update slave DoF
```
Comment (jorgensd) [inline python/demos/demo_periodic_gep.py:313]: Add blank line between above lines with only comments

File: python/demos/demo_periodic_gep.py [hunk 8]
```diff
@@ -198,82 +205,143 @@ def solve_GEP_shiftinvert(A: PETSc.Mat, B: PETSc.Mat,
...
+    for i in range(len(eigval)):
+        eigvec_r[i].x.scatter_forward()
+        mpc.backsubstitution(eigvec_r[i].vector)
+        eigvec_i[i].x.scatter_forward()
+        mpc.backsubstitution(eigvec_i[i].vector)
+    print0(f"Computed eigenvalues:\n {np.around(eigval,decimals=2)}")
+
+    # Save first eigenvector
+    with XDMFFile(MPI.COMM_WORLD, "results/eigenvector_0.xdmf", "w", encoding=XDMFFile.Encoding.HDF5) as xdmf:
```
Comment (jorgensd) [inline python/demos/demo_periodic_gep.py:322]:
```suggestion
    with XDMFFile(mesh.comm, "results/eigenvector_0.xdmf", "w") as xdmf:
```

**Review by jorgensd — CHANGES_REQUESTED** [review]
Looks alright, but the code needs to be formatted such that it passes `flake8` and `mypy` tests

File: python/demos/demo_periodic_gep.py (hunk 5 above)
Comment (fmonteghetti) [reply python/demos/demo_periodic_gep.py]: Sorry, I did not know about this requirement.
On fmonteghetti:master, flake8 does not highlight any issues.

File: python/demos/demo_periodic_gep.py (hunk 7 above)
Comment (fmonteghetti) [reply python/demos/demo_periodic_gep.py:323]: On the current fmonteghetti:master, flake8 does not highlight any issues.

**Comment by jorgensd** [comment]
@fmonteghetti Could you merge the latest version of the `main` branch into your PR?

**Comment by fmonteghetti** [comment]
> @fmonteghetti Could you merge the latest version of the `main` branch into your PR?

I have just synchronized fmonteghetti:master with jorgensd:main. Is this enough?

**Comment by jorgensd** [comment]
> > @fmonteghetti Could you merge the latest version of the `main` branch into your PR?
> 
> I have just synchronized fmonteghetti:master with jorgensd:main. Is this enough?

The CI is currently running. 
You should add your demo to the CI, so that it doesn't break in future pull requests.

**Comment by fmonteghetti** [comment]
> > > @fmonteghetti Could you merge the latest version of the `main` branch into your PR?
> > 
> > 
> > I have just synchronized fmonteghetti:master with jorgensd:main. Is this enough?
> 
> The CI is currently running. You should add your demo to the CI, so that it doesn't break in future pull requests.

Commit b2e1e6c27698365cbab7892f14e73f1e067325c4 on fmonteghetti:master updates the test_mpc.yml file.

**Review by jorgensd — APPROVED** [review]

**Pull request merged**